src/main/java/com/nourish1709/urlshortener/service/SimpleUrlShortenerService.java

```diff
@@ -0,0 +1,52 @@
+package com.nourish1709.urlshortener.service;
+
+import com.nourish1709.urlshortener.dao.ShortenedUrlRepository;
+import com.nourish1709.urlshortener.dto.ShortenUrlDto;
+import com.nourish1709.urlshortener.entity.ShortenedUrl;
+import com.nourish1709.urlshortener.exception.UrlNotFoundException;
+import lombok.RequiredArgsConstructor;
+import org.apache.commons.lang3.RandomStringUtils;
+import org.springframework.cache.annotation.Cacheable;
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+
+import java.util.Optional;
+
+@Service
+@RequiredArgsConstructor
+public class SimpleUrlShortenerService implements UrlShortenerService {
+
+    private static final int SHORTENED_URL_ID_LENGTH = 8;
+
+    private final ShortenedUrlRepository shortenedUrlRepository;
+
+    @Override
+    @Transactional
+    public ShortenedUrl shortenUrl(ShortenUrlDto shortenUrlDto) {
+        final Optional<ShortenedUrl> existingShortenedUrl = shortenedUrlRepository.findByOriginalUrl(shortenUrlDto.url());
+        if (existingShortenedUrl.isPresent()) {
+            return existingShortenedUrl.get();
+        }
+
+        final ShortenedUrl shortenedUrl = new ShortenedUrl();
+
+        String urlId = RandomStringUtils.random(SHORTENED_URL_ID_LENGTH, true, false);
+
+        while (shortenedUrlRepository.existsById(urlId)) {
```

Comment:
@nourish1709 I get that it's almost impossible to fail here, but please notice that your impl is **not thread safe.** 
Once the `while` loop is completed, **another thread can store a record with the same id.**